src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/Formatters/Binary/BinaryFormatterWriter.cs

```diff
@@ -140,55 +140,59 @@ internal void WriteObject(NameInfo nameInfo, NameInfo typeNameInfo, int numMembe
                 _binaryObject.Set(objectId, objectMapInfo._objectId);
                 _binaryObject.Write(this);
             }
-            else if (!typeNameInfo._transmitTypeOnObject)
+            else
             {
-                // ObjectWithMap
-                if (_binaryObjectWithMap == null)
+                Debug.Assert(typeNameInfo != null);
+                if (!typeNameInfo._transmitTypeOnObject)
```

Comment:
As I mentioned before `typeNameInfo` explicitly called with null, don't explicitly see this condition would not be  reachable, for now asserted it.

Reply:
Created issue https://github.com/dotnet/corefx/issues/42393